### PR TITLE
iox #27 Sample class refactoring

### PR DESCRIPTION
**Pull request by Indra5196**
## Pre-Review Checklist for the PR Author

1. [ ] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [ ] Tests follow the [best practice for testing][testing]
1. [ ] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [ ] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
    - [ ] Commit messages are signed (`git commit -s`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [ ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [ ] Relevant issues are linked
1. [ ] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes **TBD**

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+class SubscriberInterface;
+
+
+namespace internal
+{
+/// @brief helper struct for smartchunk
+template <typename TransmissionInterface, typename T, typename H>
+struct SmartChunkPrivateData
+{
+    SmartChunkPrivateData(cxx::unique_ptr<T>&& smartchunkUniquePtr, TransmissionInterface& publisher) noexcept;
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp:40]: `TransmissionInterface& publisher` sounds a bit strange. Could you please change this to `TransmissionInterface& transmitter` or `ProducerInterface& producer` or something similar which makes sense when used in the context of `Sample`, `Request` and `Response`?

File: iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    SmartChunkPrivateData(const SmartChunkPrivateData&) = delete;
+    SmartChunkPrivateData& operator=(const SmartChunkPrivateData&) = delete;
+
+    cxx::unique_ptr<const T> smartchunkUniquePtr;
+};
+} // namespace internal
+
+///
+/// @brief The SmartChunk class is a mutable abstraction over types which are written to loaned shared memory.
+/// These smartchunks are publishable to the iceoryx system.
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp:70]: Similar here with `publishable`. Wouldn't something like `transmittable`/`transferable` be more generic and also fit to request-response?

File: iceoryx_posh/include/iceoryx_posh/popo/sample.hpp [hunk 3]
```diff
@@ -29,58 +27,23 @@ namespace popo
...
+class Sample : public SmartChunk<PublisherInterface<T, H>, T, H>
 {
     static_assert(std::is_const<T>::value == std::is_const<H>::value,
                   "The type `T` and the user-header `H` must be equal in their const qualifier to ensure the same "
                   "access restrictions for the user-header as for the sample data!");
 
-    /// @brief Helper type to enable the constructor for the publisher, i.e. when T has no const qualifier
     template <typename S, typename TT>
-    using ForPublisherOnly = std::enable_if_t<std::is_same<S, TT>::value && !std::is_const<TT>::value, S>;
+    using ForProducerOnly = std::enable_if_t<std::is_same<S, TT>::value && !std::is_const<TT>::value, S>;
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/popo/sample.hpp:42]: I did not test it, bud doesn't this also work?
```cpp
using SmartChunk<PublisherInterface<T, H>, T, H>::ForProducerOnly;
```

File: iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp:2]: If you want, you can also add your copyright here

**Review by elBoberido — COMMENTED** [review]
Reviewed only `SmartChunk` and `Sample` for now.

Please link to the related issue. This can be done by adding `Relates to #27` at the end of the PR description.

Did you run clang-format? If not, we recommend to use the git-hooks from `tools/git-hooks`. There is also a readme with the recommended installation procedure in that folder

File: iceoryx_posh/include/iceoryx_posh/popo/sample.hpp (hunk 3 above)
Comment (Indra5196) [reply iceoryx_posh/include/iceoryx_posh/popo/sample.hpp:42]: No it does not. Dont know the exact reason though.

File: iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp (hunk 1 above)
Comment (Indra5196) [reply iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp]: Done

File: iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp (hunk 2 above)
Comment (Indra5196) [reply iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp]: Done

File: iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp (hunk 4 above)
Comment (Indra5196) [reply iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp:2]: Done

File: iceoryx_posh/include/iceoryx_posh/popo/sample.hpp (hunk 3 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/popo/sample.hpp:44]: I think the following should work
```cpp
using Base_t = SmartChunk<PublisherInterface<T, H>, T, H>;

template <typename S, typename TT>
using ForProducerOnly = typename Base_t::template ForProducerOnly<S, TT>;
```

**Comment by elBoberido** [comment]
For traceability, we use commit messages like `iox-#27 Commit message`. Please adjust the commit messages.

It is similar with the branches. We use `iox-#27-branch-name` for branch names. I think this cannot be changed after the PR is created. You might need to close this and crate new one.

**Comment by Indra5196** [comment]
> For traceability, we use commit messages like `iox-#27 Commit message`. Please adjust the commit messages.
> 
> It is similar with the branches. We use `iox-#27-branch-name` for branch names. I think this cannot be changed after the PR is created. You might need to close this and crate new one.

**Pull request closed without merging**